pacman/operations/routing_info_allocator_algorithms/zoned_routing_info_allocator.py

```diff
@@ -16,154 +16,309 @@
+
+    def __check_zones(self):
+        # See if it could fit even before considerding fixed
+        app_part_bits = self.__bits_needed(len(self.__atom_bits_per_app_part))
+        if app_part_bits + self.__n_bits_atoms_and_mac > BITS_IN_KEY:
             raise PacmanRouteInfoAllocationException(
                 "Unable to use ZonedRoutingInfoAllocator please select a "
                 "different allocator as it needs {} + {} bits".format(
-                    source_bits, max_app_keys_bits))
-        if max_partitions > self.MAX_PARTITIONS_SUPPORTED:
-            raise NotImplementedError()
-        return max_app_keys_bits, key_bits_per_app
+                    app_part_bits, self.__n_bits_atoms_and_mac))
 
-    def _simple_allocate(self, max_app_keys_bits, key_bits_map):
-        """
-        :param int max_app_keys_bits: max bits for app keys
-        :param dict(ApplicationVertex,int) key_bits_map:
-            map of app vertex to max keys for that vertex
-        :return: tuple of routing infos and map from app vertex and key masks
-        :rtype: tuple(RoutingInfo, dict(ApplicationVertex,BaseKeyAndMask))
-        """
+        # Reserve fixed and check it still works
+        self.__set_fixed_used()
+        app_part_bits = self.__bits_needed(
+            len(self.__atom_bits_per_app_part) + len(self.__fixed_used))
+        if app_part_bits + self.__n_bits_atoms_and_mac > BITS_IN_KEY:
+            raise PacmanRouteInfoAllocationException(
+                "Unable to use ZonedRoutingInfoAllocator please select a "
+                "different allocator as it needs {} + {} bits".format(
+                    app_part_bits, self.__n_bits_atoms_and_mac))
+
+        if not self.__flexible:
+            if app_part_bits + self.__n_bits_machine + self.__n_bits_atoms > \
+                    BITS_IN_KEY:
+                # We know from above test that all will fit if flexible
+                # Reduce the suggested size of n_bits_atoms
+                self.__n_bits_atoms = \
+                    BITS_IN_KEY - app_part_bits - self.__n_bits_machine
+                self.__n_bits_atoms_and_mac = BITS_IN_KEY - app_part_bits
+            else:
+                # Set the size of atoms and machine for biggest of each
+                self.__n_bits_atoms_and_mac = \
+                    self.__n_bits_machine + self.__n_bits_atoms
+
+    def __check_flexible_zones(self):
+        if self.__n_bits_atoms_and_mac == \
+                self.__n_bits_machine + self.__n_bits_atoms:
+            return  # felibible same as none flexible
+
+        raw_app_part_bits = self.__bits_needed(
+            len(self.__atom_bits_per_app_part))
+
+        while self.__n_bits_atoms >= 0:
+            self.__n_bits_atoms_and_mac = \
+                self.__n_bits_machine + self.__n_bits_atoms
+            if self.__n_bits_atoms_and_mac + raw_app_part_bits <= BITS_IN_KEY:
+                self.__set_fixed_used()
```

Comment:
Now I see this is being used, a couple of extra comments.  Firstly, this seems to be called twice.  I can't see it doing anything different the second time, but I could be wrong.

Reply:
The second call was well OOPS!  Good spot thanks!